docs/src/content/docs/book/assembly-functions.mdx

```diff
@@ -0,0 +1,678 @@
+struct Nested { ab1: AB; ab2: AB }
+
+// This will multiply the sums of fields of nested `AB` Structs
+asm fun mulOfSums(n: Nested): Int { ADD -ROT ADD MUL }
+
+// Action!
+fun showcase() {
+    sum(AB{ a: 27, b: 50 }); // 77
+    //      ↑      ↑
+    //      |      Pushed last, sits on top of the stack
+    //      Pushed first, sits on the bottom of the stack
+
+    mulOfSums(Nested{ ab1: AB{ a: 1, b: 2 }, ab2: AB{ a: 3, b: 4 } }); // 21
+    //                         ↑     ↑                ↑     ↑
+    //                         |     |                |     Pushed last,
+    //                         |     |                |     sits on top of the stack
+    //                         |     |                Pushed second-to-last,
+    //                         |     |                sits below the top of the stack
+    //                         |     Pushed second,
+    //                         |     sits right above the bottom of the stack
+    //                         Pushed first, sits on the bottom of the stack
+}
+```
+
+### Returns {#conventions-returns}
+
+When present, return type of an assembly function attempts to capture relevant values from the resulting stack after the function execution and possible stack [arrangements](#arrangements). When not present, however, assembly function does not take any values from the stack.
+
+When present, an assembly function's return type attempts to grab relevant values from the resulting stack after the function execution and any [result arrangements](#arrangements). If the return type is not present, however, the assembly function does not take any values from the stack.
+
+```tact
+// Pushes `x` onto the stack, increments it there,
+// but does not capture the result, leaving it on the stack
+asm fun push(x: Int) { INC }
+```
+
+Specifying a [primitive type][p], such as an [`Int{:tact}`][int] or a [`Cell{:tact}`][cell], will make the assembly function pop the top value from the stack and produce it as a result. If the run-time type of the popped value doesn't match the specified return type, an exception with [exit code 7](/book/exit-codes#7) will be thrown: `Type check error`.
+
+```tact
+// CAUSES RUN-TIME ERROR!
+// Pushes `x` onto the stack, then tries to capture it as a Cell,
+// causing an exit code 7: Type check error
+asm fun push(x: Int): Cell { }
+```
+
+Just like in [parameters](#conventions-parameters), arbitrary [Struct][struct] return types are distributed across their fields and recursively flattened in exactly the same order. The only differences are that they now capture or pop values from the stack instead of pushing them onto the stack, and they do so in a right-to-left fashion — the last field of the [Struct][struct] pops the topmost value from the stack, the second-to-last pops the second to the top, and so on, so that the last field contains the value from the top of the stack and the first field contains the value from the bottom.
```

Comment:
popping here is confusing ("pops" = "represents"?)

also, are we sure about flattening here?

Reply:
> also, are we sure about flattening here?

Wdym? It's the current behavior. Do we not want to capture multiple values from the stack at once, individually, and not as a tuple? That'd be detrimental to power-users with zero benefit to regular users

Reply:
ok, the flattening question has been resolved; let's figure out what to do with the "pops" terminology here

Reply:
I've removed that :)